### PR TITLE
python3Packages.manuf: init at 1.1.5

### PR DESCRIPTION
**Pull request by dani0854**
## Description of changes

Added a python package manuf

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dani0854** [comment]
Result of `nixpkgs-review pr 255080` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.manuf</li>
    <li>python310Packages.manuf.dist</li>
    <li>python311Packages.manuf</li>
    <li>python311Packages.manuf.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/manuf/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, runCommand
+, python3
+}:
+
+buildPythonPackage rec {
+  pname = "manuf";
+  version = "1.1.5";
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/default.nix:10]:
```suggestion
  version = "1.1.5";
  format = "setuptools";
```
Would you please add `format`?

File: pkgs/development/python-modules/manuf/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  src = fetchFromGitHub {
+    owner = "coolbho3k";
+    repo = "manuf";
+    rev = "${version}";
+    sha256 = "sha256-3CFs3aqwE8rZPwU1QBqAGxNHT5jg7ymG12yBD56gTNI=";
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/default.nix:16]:
```suggestion
    hash = "sha256-3CFs3aqwE8rZPwU1QBqAGxNHT5jg7ymG12yBD56gTNI=";
```

File: pkgs/development/python-modules/manuf/default.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  };
+
+  # Tests appeared to be broken
+  doCheck = false;
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/default.nix:20]: Why doesn't the test work?
Is it because of the sandbox environment?

File: pkgs/development/python-modules/manuf/default.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      [ "$(cat $out | tr -d '\n')" = "Vendor(manuf='ASUSTekC', manuf_long='ASUSTek COMPUTER INC.', comment=None)" ]
+    '';
+  };
+
+  meta = with lib; {
+    homepage = "https://github.com/coolbho3k/manuf";
+    description = " Parser library for Wireshark's OUI database";
+    platforms = platforms.linux;
+    license = with licenses; [ lgpl3Plus asl20 ];
+    maintainers = with maintainers; [ dsuetin ];
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/default.nix:34]:
```suggestion
    maintainers = with maintainers; [ dsuetin ];
    mainProgram = "manuf";
```
I think this may be useful.

File: pkgs/development/python-modules/manuf/default.nix [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  # Tests appeared to be broken
+  doCheck = false;
+
+  passthru.tests = {
+    testMacAddress = runCommand "${pname}-test" {} ''
+      ${python3.pkgs.manuf}/bin/manuf BC:EE:7B:00:00:00 > $out
+      [ "$(cat $out | tr -d '\n')" = "Vendor(manuf='ASUSTekC', manuf_long='ASUSTek COMPUTER INC.', comment=None)" ]
+    '';
+  };
+
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/default.nix:28]: Could you please enable pythonImportsCheck for the simple check?

File: pkgs/development/python-modules/manuf/default.nix (hunk 3 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix:20]: While executing test via nix build I get
```
Traceback (most recent call last):
  File "/build/source/manuf/manuf.py", line 278, in _get_mac_int
    return int(mac_str, 16) << self._bits_left(mac_str)
ValueError: invalid literal for int() with base 16: 'test'
```

Which is most likely due to an incorrect way of launching tests, and can be fixed.

But even if it is fixed, there are still test that fail in general https://github.com/coolbho3k/manuf/actions. Update functionality seem to be broken, and is not required due to the nix way of updating packages. Of course, it is possible to patch tests, removing the broken one. But I thought nix test will be sufficient in that case, since this package is quite simple. 

I have extended comment why the tests has been turned off, but you think it is important to get them working, I can try doing that.

File: pkgs/development/python-modules/manuf/default.nix (hunk 5 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix:28]: I don't think that the main python file imports anything outside the standard library. Of course, I can make a check that standard library has all imports required if it is necessary.

Although tests have some required packages, but they are turned off.

File: pkgs/development/python-modules/manuf/default.nix (hunk 1 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix:10]: Done

File: pkgs/development/python-modules/manuf/default.nix (hunk 2 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix]: Done

File: pkgs/development/python-modules/manuf/default.nix (hunk 4 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix:39]: Done

File: pkgs/development/python-modules/manuf/default.nix (hunk 5 above)
Comment (natsukium) [reply pkgs/development/python-modules/manuf/default.nix:32]: We can find that its dependencies are all standard libraries, but there are many packages that import implicit dependencies such as `pkg_resources` or `importlib_metadata`.
Adding this check would be a good choice since it has little overhead.

Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix:34]: Added `pythonImportsCheck`. But not sure if I did it correctly.

File: pkgs/development/python-modules/manuf/default.nix (hunk 3 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix]: Turns out manuf doesn't work without updating it's database in some cases anyway. So will have to fix that nevertheless.

Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix]: Ok, added a patch, tests working now

**Comment by dani0854** [comment]
Result of `nixpkgs-review pr 255080` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.manuf</li>
    <li>python310Packages.manuf.dist</li>
    <li>python311Packages.manuf</li>
    <li>python311Packages.manuf.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/manuf/default.nix [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  src = fetchFromGitHub {
+    owner = "coolbho3k";
+    repo = "manuf";
+    rev = "${version}";
+    hash = "sha256-3CFs3aqwE8rZPwU1QBqAGxNHT5jg7ymG12yBD56gTNI=";
+  };
+
+  nativeBuildInputs = [ wireshark-cli ];
+
+  prePatch = ''
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/default.nix:23]:
```suggestion
  postPatch = ''
```
prePatch is really only used for operations that need to be done before the patch phase, such as dos2unix, which converts the file format.

File: pkgs/development/python-modules/manuf/default.nix [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    cat ${wireshark-cli}/share/wireshark/{manuf,wka} > manuf/manuf
+  '';
+
+  patches = [ ./internal_db_update_nix.patch ];
+
+  checkPhase = ''
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/default.nix:29]: I think we can use `pytestCheckHook` here.

File: pkgs/development/python-modules/manuf/internal_db_update_nix.patch [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+index e5e9193..0ac9296 100755
+--- a/manuf/manuf.py
++++ b/manuf/manuf.py
+@@ -61,12 +61,19 @@ class MacParser(object):
+         IOError: If manuf file could not be found.
+ 
+     """
+-    MANUF_URL = "https://gitlab.com/wireshark/wireshark/raw/master/manuf"
++    # https://github.com/coolbho3k/manuf/issues/34
++    MANUF_URL = "https://www.wireshark.org/download/automated/data/manuf"
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/internal_db_update_nix.patch:11]: Is this related to the db update?
If not, it would be easier to maintain if we split it into separate patches.

File: pkgs/development/python-modules/manuf/internal_db_update_nix.patch [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+index b60ffe2..c5816ff 100644
+--- a/manuf/test/test_manuf.py
++++ b/manuf/test/test_manuf.py
+@@ -8,7 +8,8 @@ class ManufTestCase(unittest.TestCase):
+ 
+     def setUp(self):
+         self.manuf = manuf.MacParser(manuf_name="manuf/test/manuf")
+-    #
++
++    @unittest.skip("Won't work on nix, requires internet access")
```
Comment (natsukium) [inline pkgs/development/python-modules/manuf/internal_db_update_nix.patch:48]: We can use `disabledTests` with `pytestCheckHook`.

**Comment by natsukium** [comment]
Could you please change the PR title and commit message from `pythonPackages` to `python310Packages`?
pythonPackages is a package of Python2 currently.

File: pkgs/development/python-modules/manuf/default.nix (hunk 6 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix]: Done

File: pkgs/development/python-modules/manuf/default.nix (hunk 7 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/default.nix]: Done

File: pkgs/development/python-modules/manuf/internal_db_update_nix.patch (hunk 8 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/internal_db_update_nix.patch]: Done

File: pkgs/development/python-modules/manuf/internal_db_update_nix.patch (hunk 9 above)
Comment (dani0854) [reply pkgs/development/python-modules/manuf/internal_db_update_nix.patch]: Done

**Comment by dani0854** [comment]
> Could you please change the PR title and commit message from `pythonPackages` to `python310Packages`? pythonPackages is a package of Python2 currently.

Changed to `python3Packages`, since it builds for both `python310Packages` and `python311Packages`

**Comment by dani0854** [comment]
Result of `nixpkgs-review pr 255080` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.manuf</li>
    <li>python310Packages.manuf.dist</li>
    <li>python311Packages.manuf</li>
    <li>python311Packages.manuf.dist</li>
  </ul>
</details>

**Comment by natsukium** [comment]
@ofborg build python310Packages.manuf python311Pacakges.manuf

**Review by natsukium — APPROVED** [review]
LGTM
